src/Compiler/Driver/ReuseTcResults/ReuseTcResultsDriver.fs

```diff
@@ -0,0 +1,124 @@
+module internal FSharp.Compiler.ReuseTcResults
+
+open System
+open System.Collections.Generic
+open System.IO
+
+open FSharp.Compiler.CompilerConfig
+open FSharp.Compiler.Diagnostics
+open FSharp.Compiler.GraphChecking
+open FSharp.Compiler.IO
+open FSharp.Compiler.Syntax
+open FSharp.Compiler.Syntax.PrettyNaming
+
+open Internal.Utilities.Hashing
+
+let (++) a b = Path.Combine(a, b)
+
+[<Sealed>]
+type ReuseTcResultsDriver(tcConfig: TcConfig) =
+
+    let outputDir = tcConfig.outputDir |> Option.defaultValue ""
+    let tcDataFolderName = outputDir ++ FSharpTcDataResourceName
+
+    let cmdLineFilePath = tcDataFolderName ++ "cmdline"
+    let graphFilePath = tcDataFolderName ++ "graph"
+    let referencesFilePath = tcDataFolderName ++ "references"
+
+    let writeThisTcData path content =
+        use tcDataFile = FileSystem.OpenFileForWriteShim path
+        tcDataFile.WriteAllLines content
+
+    let readPrevTcData path =
+        if FileSystem.FileExistsShim path then
+            use tcDataFile = FileSystem.OpenFileForReadShim path
+            tcDataFile.ReadAllLines()
+        else
+            Array.empty
+
+    let getPrevCompilationCmdLine () = readPrevTcData cmdLineFilePath
+    let getPrevCompilationGraph () = readPrevTcData graphFilePath
+    let getPrevCompilationReferences () = readPrevTcData referencesFilePath
+
+    let writeThisCompilationCmdLine = writeThisTcData cmdLineFilePath
+    let writeThisCompilationGraph = writeThisTcData graphFilePath
+    let writeThisCompilationReferences = writeThisTcData referencesFilePath
```

Comment:
All writers should use buffer and flush it int the very end to avoid partial data being in the cached file.

Reply:
Now there is only one write operation, does this still apply?
If so, any inspiration in our code?
I am not sure I understand - I would imagine our file system shim to do all this important stuff.